### ISSUE TITLE
False Positive caused by Angular2 - ng-reflect?

### ISSUE THREAD

**Issue by BlakSwan** [dequelabs/axe-core#235]
I am working on adding accessibility to my company's new site developed in Angular2.  We have come across a violation in aXe that has us puzzled.  The violation message is as follows:

Document has multiple elements with the same id attribute: ps-radio-1

The source code of the violation is:

<input class="ps-radio__control" type="radio" ng-reflect-id="ps-radio-1" id="ps-radio-1" ng-reflect-checked="false" ng-reflect-disabled="false">

As you can see the violation is caused by Angular and the ng-reflect attribute.  Is this something that will cause accessibility issues? Or is this a false violation?

Thanks so much

**Comment by marcysutton** [comment]
Hi @BlakSwan, is it possible to include a gist, jsbin or Plunker showing the issue? It will help us get to a fix much faster.

**Comment by BlakSwan** [comment]
Sorry, this is what the compiled code looks like, sorry it didn't show up in the first one.

`<input class="ps-radio__control" type="radio" ng-reflect-id="ps-radio-1" id="ps-radio-1" ng-reflect-checked="false" ng-reflect-disabled="false">`



Here is the component template:
`<label class="ps-radio__label">
  <div class="ps-radio__container">
    <div class="ps-radio__outer-circle"></div>
    <div class="ps-radio__inner-circle"></div>
  </div>

  <input
    class="ps-radio__control"
    #radio
    type="radio"
    [id]="id"
    [attr.name]="name"
    [attr.tabindex]="tabIndex"
    [attr.aria-label]="ariaLabel"
    [attr.aria-labelledby]="ariaLabelledby"
    [checked]="checked"
    [disabled]="disabled"
    (blur)="_handleBlur($event)"
    (change)="_handleChange($event)"
    (focus)="_handleFocus($event)"
  >
  <span class="ps-radio__label-content">{{label}}</span>
</label>
`

**Comment by dylanb** [comment]
@BlakSwan This will definitely cause accessibility issues that could be blockers for certain types of users.

For example, when a Dragon voice recognition user wants to interact with a radio button, they would say "Click Yes" (assuming that the radio button's associated label's text is "Yes"), Dragon would then look for a clickable element with the label of "Yes", if it finds only one, it extracts the ID from that element and sends a click event to the element with that id.

If there are multiple elements on the page with the same id, the first element in the DOM is the one that will receive the click event. This could result in the wrong thing getting clicked, or it could result in nothing happening at all (if the id is on a non-interactive element in addition to the interactive element).

This can result in functionality being totally unusable for the user of that assistive technology.

In addition to that, it is non-conforming HTML, so a good developer should avoid it :-)

**Comment by BlakSwan** [comment]
Hi @dylanb Thanks for the detailed explanation!  That is what I thought was going to happen...the question is what can be done in Angular2 to stop this from happening?  The ng-reflect attributes are system driven at compile, not something I have added.

**Comment by dylanb** [comment]
Is it not only a debug-time feature?

**Comment by BlakSwan** [comment]
Also the attributes have two different names so not sure why they are both showing as id's.  I am 100% sure it is not a debug feature

**Comment by dylanb** [comment]
@BlakSwan a plunker would definitely help

**Comment by BlakSwan** [comment]
@dylanb I will try and put one together on Monday.  Just to check I have built my project in production mode via three different methods (gulp, angular-cli, and enableProdMode in the bootstrap file).  They all have ng-reflect attributes still.  I will get back to you on Monday.

Thanks

**Comment by BlakSwan** [comment]
Hello @dylanb and @marcysutton, sorry it took a while to finish the plunker.  Here it is:  

https://plnkr.co/edit/wTPfUdDMAGdgkVUSeu4p?p=preview

I have switched it to production mode and have confirmed that Dylan was correct, the ng-reflect is not created in this mode.  However, I still wonder, how are we supposed to check for accessibility during development if the aXe pluggin does not recognize the difference between the id attribute and the ng-reflect-id attribute?

Thanks

**Comment by marcysutton** [comment]
Now that I can see the rendered source, I can actually comment:

There are two IDs in the DOM, so aXe is finding a legitimate issue (see `id="ps-radio-1"`). That seems like an issue with Angular 2.

```
<ps-radio label="Male" value="male" class="ps-radio--is-inline" id="ps-radio-1"><label class="ps-radio__label">
  <div class="ps-radio__container">
    <div class="ps-radio__outer-circle"></div>
    <div class="ps-radio__inner-circle"></div>
  </div>

  <input class="ps-radio__control" type="radio" id="ps-radio-1">
  <span class="ps-radio__label-content">Male</span>
</label></ps-radio>
```

**Comment by BlakSwan** [comment]
Hi Marcy,  thanks for your quick response.  The second id you are showing is on a different class and different element type ie an input and a ps-radio.  Do the screen readers not ignore the Angular 2 created elements?

Thanks

**Comment by marcysutton** [comment]
Screen readers, as well as aXe, only care about what's in the rendered DOM source, which in this case includes the same ID repeated more than once. It doesn't matter that they are different classes or element types, or that they were created with Angular 2–a tree structure with multiple of the same ID is invalid HTML (which says ids must be unique). This can cause problems when pairing a label with an input because it can only really apply one of them–and as a testing library, we can't reliably infer which IDREF is the correct one based on source order. So we flag it, because it's a bad practice.

There was a discussion about this in a11ySlackers the other day, it might be useful to you: http://w3c.github.io/a11ySlackers/archives/2016/10/2016-10-04.html#chat-item-15
